### PR TITLE
feat: Add support for views with Lens transforms

### LINKED ISSUE #2070: Add support for un-cached Lens-driven Defra Views
Add support for un-cached Lens-driven Defra Views.

Persistence/caching of the result will be added later. Mutations through the view (to underlying documents) will be added later. P2P stuff will be added later.

### PR DESCRIPTION
**Pull request by AndrewSisley**
## Relevant issue(s)

Resolves #2070

## Description

Adds support for views with Lens transforms.

### CODE REVIEW

File: tests/lenses/rust_wasm32_filter/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+static PARAMETERS: RwLock<StreamOption<Parameters>> = RwLock::new(None);
+
+#[no_mangle]
+pub extern fn alloc(size: usize) -> *mut u8 {
+    lens_sdk::alloc(size)
+}
+
+#[no_mangle]
```
Comment (fredcarle) [inline tests/lenses/rust_wasm32_filter/src/lib.rs:49]: what is no_mangle

**Review by fredcarle — APPROVED** [review]
LGTM. I've create a few follow up issues that stemmed from reviewing this.

- https://github.com/sourcenetwork/defradb/issues/2314
- https://github.com/sourcenetwork/defradb/issues/2315
- https://github.com/sourcenetwork/defradb/issues/2316

File: planner/lens.go
Comment (shahzadlone) [inline planner/lens.go:1]: todo: Since a new plan node is being added, I would make the same comment as: https://github.com/sourcenetwork/defradb/pull/2114#discussion_r1432460060

tldr: Please add the new name to `allPlanNodeNames` variable in the `tests/integration/explain.go` file for debug node and add a test for the debug explain under the folder `tests/integration/explain/debug` similar to `tests/integration/explain/debug/with_view_test.go`. You don't have to make it explainable however in this PR, that discussion can be had elsewhere but would suggest opening an issue for it.

**Review by shahzadlone — CHANGES_REQUESTED** [review]
Mostly looks good, just the one main concern regarding the explain debug test similar to this comment when view node was added: https://github.com/sourcenetwork/defradb/pull/2114#discussion_r1432460060

**Comment by shahzadlone** [comment]
> LGTM. I've create a few follow up issues that stemmed from reviewing this.
> 
> * [Add explain method to lens node #2314](https://github.com/sourcenetwork/defradb/issues/2314)
> * [Add explain method to view node #2315](https://github.com/sourcenetwork/defradb/issues/2315)
> * [Add support for foreign objects in document creation. #2316](https://github.com/sourcenetwork/defradb/issues/2316)

Ahh cheers for making the explain issues. Are the issues for discussion if we need lens and view nodes, or we already decided that we need explain for them?

File: planner/lens.go
Comment (AndrewSisley) [reply planner/lens.go:1]: - [ ] Add explain test stuff

**Comment by AndrewSisley** [comment]
> Are the issues for discussion if we need lens and view nodes, or we already decided that we need explain for them?

I don't think that they needing explain is a given, although I think the fact that we have this question and hide some nodes from our users is an unwanted complication for both us and them and plan on chatting about this in the standup (it came up in some of my other work).

**Comment by AndrewSisley** [comment]
> LGTM. I've create a few follow up issues that stemmed from reviewing this.

Cheers Fred :)

File: planner/lens.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+func (n *lensNode) Start() error {
+	return n.source.Start()
+}
+
+func (n *lensNode) Spans(spans core.Spans) {
+	n.source.Spans(spans)
+}
+
+func (n *lensNode) Next() (bool, error) {
+	hasNext, err := n.output.Next()
```
Comment (islamaliev) [inline planner/lens.go:69]: nitpick: var with name `hasValue` would fit better here

File: planner/lens.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		// similar logic, for example when converting a core.Doc to a map before passing it
+		// into a lens transform.
+		fieldIndex := indexes[len(indexes)-1]
+
+		if len(properties) <= fieldIndex {
+			// Because the document is sourced from another mapping, we may still need to grow
+			// the resultant field set. We cannot use [append] because the index of each field
+			// must still correspond to it's field ID.
+			originalProps := properties
+			properties = make([]any, fieldIndex+1)
```
Comment (islamaliev) [inline planner/lens.go:144]: suggestion: it will be more efficient to run through `indexes` to find the maximum value and then allocate `properties`. It can be also done before the loop.
Otherwise there might be scenarios when `properties` can be allocated several times.

Comment (AndrewSisley) [reply planner/lens.go:144]: Which one is more efficient depends on the two sizes really.  Not every indexes key will be present in the map and your suggestion will over-allocate when that happens.

Your suggestion is probably better more of the time, but I don't think it is worth a new PR.
